### PR TITLE
[CI:DOCS] ci(lint): Add Shell linter - Differential Shellcheck

### PR DESCRIPTION
**Pull request by jamacku**
Differential ShellCheck is a GitHub action that performs differential ShellCheck scans on shell scripts changed via PR and reports results directly in PR. The beauty of differential scans is that you will get reports only about newly added defects. 

Since this repository has few shell scripts, I think you might find some value in having a shell linter. Differential ShellCheck action is able to produce reports in SARIF format. GitHub understands this format and is able to display it nicely as a PR comment, and on the `Files Changed` tab, please see below.

image

image

Documentation is available at [@redhat-plumbers-in-action/differential-shellcheck](https://github.com/redhat-plumbers-in-action/differential-shellcheck#readme). Let me know If you are missing some feature or setting. I'm always happy to extend functionality.

> **Note**: I have added the ShellCheck directive to all `*.at` files, so ShellCheck would be able to detect these files as shell scripts correctly. Also, please ignore reported errors; the added directives cause them. When this PR is merged, these errors won't be reported again.

---

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by rhatdan** [comment]
@edsantiago @cevich PTAL

File: test/apiv2/01-basic.at
Comment (edsantiago) [inline test/apiv2/01-basic.at:1]: Please leave the `-*-` line first: this is needed by [emacs](https://www.gnu.org/software/emacs/manual/html_node/emacs/Choosing-Modes.html) to determine syntax highlighting. If this new tool supports it, you could do:
```
# -*- sh -*- shellcheck shell=sh
```

**Review by edsantiago — CHANGES_REQUESTED** [review]
Interesting idea! My immediate response is, can we see the results of this *without* differential check, to get a feeling for what sorts of issues it will flag, and especially what the false-alarms will be? I'm not eager to enable a new check without first seeing how it reacts to our existing code.

No hurry: I'm OOTO today, and Monday is a holiday, so the earliest I can look would be Tuesday.

Thank you!

File: test/apiv2/01-basic.at
Comment (jamacku) [reply test/apiv2/01-basic.at:1]: Yeah, I felt that it needs to be at the first line. Unfortunately, your suggested solution would not work either. Shellcheck requires that its directives are at the beginning of the line.

But I think it would be helpful to add support for emacs filetypes to `differential-shellcheck`:
- https://github.com/redhat-plumbers-in-action/differential-shellcheck/issues/101

The solution could look something like this:

```sh
# -*- sh -*-
# shellcheck shell=sh
#
# Tests for more image-related endpoints
#
```

**Comment by TomSweeneyRedHat** [comment]
Once the conversation with @edsantiago is decided upon, the rest LGTM

File: .github/workflows/differential-shellcheck.yml
Comment (cevich) [inline .github/workflows/differential-shellcheck.yml:14]: Ugh, the documentation is horrible.  Best I could find is https://docs.github.com/en/rest/overview/permissions-required-for-github-apps#permission-on-security-events

I'm wondering why security-events ***write*** is needed for this workflow.  Seems the default (read) should be good enough, no?

Comment (jamacku) [reply .github/workflows/differential-shellcheck.yml:14]: Yeah, GitHub documentation is sometimes confusing and even incomplete. Documentation for token permissions is here:

https://docs.github.com/en/rest/code-scanning#upload-an-analysis-as-sarif-data

> Uploads SARIF data containing the results of a code scanning analysis to make the results available in a repository. You must use an access token with the security_events scope to use this endpoint for private repositories. You can also use tokens with the public_repo scope for public repositories only. GitHub Apps must have the security_events write permission to use this endpoint.

It is required to have `security-events: write` to successfully "write" the SARIF report to GitHub for further processing.

You can also check other GitHub actions that upload SARIF (e.g. https://github.com/github/codeql-action#usage). They also require `security-events: write`

I hope this answers your question.

File: test/apiv2/01-basic.at
Comment (jamacku) [reply test/apiv2/01-basic.at]: I had added support for emacs file type specifications, so `differential-shellcheck` now can detect files with `# -*- sh -*-`. So there is no need for adding `# shellcheck shell=sh`.

But it might still be valuable to add a ShellCheck directive to specify if the files are supposed to be POSIX shell or bash.

**Comment by jamacku** [comment]
> Interesting idea! My immediate response is, can we see the results of this _without_ differential check, to get a feeling for what sorts of issues it will flag, and especially what the false-alarms will be? I'm not eager to enable a new check without first seeing how it reacts to our existing code.

I merged this PR into my fork `main` branch. And created testing PR:
- https://github.com/jamacku/podman/pull/3

Feel free to open PR against my fork if you want to test it yourself. If you want to see an example of the repository where is `differential-shellcheck` used, I would suggest looking at [systemd](https://github.com/systemd/systemd/actions/workflows/differential-shellcheck.yml).

There should be no false alarms. `differential-shellcheck` reports only issues caused by proposed PR changes in PRs. It is designed for repositories with existing shell code, so they could introduce shellcheck linter without fixing or masking all warnings beforehand.

Feel free to ask. I'm happy to provide more context and examples.

File: .github/workflows/differential-shellcheck.yml
Comment (cevich) [reply .github/workflows/differential-shellcheck.yml:14]: It does, but it seems likely others may also question this.  People see "security" and "write" and tend to freak-out first, but then Github doesn't make it easy for them to figure it out either.  Maybe putting that URL into a comment above the permission would help alleviate some fears?

**Review by cevich — APPROVED** [review]
LGTM with just the one security-comment addition desired.  FYI, ***before*** you force-push, edit the PR title and add `[CI:DOCS]` to it.  There's no need to make you (or anybody) wait for the full Cirrus-CI run here since it's only a github action you're touching :smile:

File: .github/workflows/differential-shellcheck.yml
Comment (jamacku) [inline .github/workflows/differential-shellcheck.yml:7]: I have noticed branches related to versions. Would you like to run a linter also against those branches?

**Review by jamacku — COMMENTED** [review]
I have added comments and links to documentation, @cevich PTAL

File: .github/workflows/differential-shellcheck.yml
Comment (rhatdan) [reply .github/workflows/differential-shellcheck.yml:7]: I don't think that would be a good idea.

Comment (cevich) [reply .github/workflows/differential-shellcheck.yml:7]: Ya, no need for that, it makes CI too complicated on the brain :grin:

**Comment by rhatdan** [comment]
@edsantiago This one is waiting on you.

**Review by edsantiago — APPROVED** [review]

**Comment by rhatdan** [comment]
/lgtm
Thanks @jamacku

**Comment by edsantiago** [comment]
Argh - sorry, too many distractions this week.

File: .github/workflows/differential-shellcheck.yml
Comment (jamacku) [reply .github/workflows/differential-shellcheck.yml]: So let's run the linter only on `main`. Thanks for your feedback. :+1:

**Comment by edsantiago** [comment]
I never saw an answer to my original question (full results), so I can't actually LGTM, but will let others do so if they wish.

**Comment by jamacku** [comment]
> I never saw an answer to my original question (full results), so I can't actually LGTM, but will let others do so if they wish.

@edsantiago By _full results_, do you mean all ShellCheck warnings on podman repository?

**Comment by edsantiago** [comment]
Yes: I was hoping to see what sort of issues this triggers on, because that will give me a sense for what problems we can expect to see. If we currently have a hundred instances of `foo bar` that this tool objects to, and someone copy/pastes a new one, they will be surprised/annoyed to find that their copy/paste fails lint.

**Comment by jamacku** [comment]
> Yes: I was hoping to see what sort of issues this triggers on, because that will give me a sense for what problems we can expect to see. If we currently have a hundred instances of `foo bar` that this tool objects to, and someone copy/pastes a new one, they will be surprised/annoyed to find that their copy/paste fails lint.

That is a 100% valid concern. I'll prepare a report of all ShellCheck warnings, but probably not today. I'm going to bed soon :smile:. I'd post it here tomorrow morning. Thanks for your patience.

**Comment by cevich** [comment]
> I'll prepare a report of all ShellCheck warnings, but probably not today. 

We've already got the git post-commit hook running for the validate job.  So seeing the full thing is as easy as:

```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index ca023d18a..128b96c53 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -18,3 +18,14 @@ repos:
       - id: check-executables-have-shebangs
       - id: check-merge-conflict
       - id: check-yaml
+      - id: shellcheck
+        args:
+            - --color=always
+            - --format=tty
+            - --shell=bash
+            - --external-sources
+            - --enable=add-default-case,avoid-nullary-conditions,check-unassigned-uppercase
+            - --exclude=SC2046,SC2034,SC2090,SC2064
+            - --wiki-link-count=0
+            - --severity=warning
+
```

Or something similar.  In fact, this may be a better alternative to using github-actions and a differential check.  Then we can just get everything actually fixed :smile: 

(I'm willing to help do fixes if needed)

**Comment by cevich** [comment]
Clarification:  `make validate` depends on `make lint` which will run `pre-commit run -a` on our CI VMs.  Or you can use `make install.pre-commit` to install it locally and run it.

**Comment by jamacku** [comment]
@edsantiago As promised, here is a list of ShellCheck defects: 
- https://gist.github.com/jamacku/cad94166aaae6807aa2e0beafb69ed65

Summary:
**188** shell scripts
**899** defects

Command used:

```sh
shellcheck \
  --exclude=SC2046,SC2034,SC2090,SC2064 \
  --enable=add-default-case,avoid-nullary-conditions,check-unassigned-uppercase \
  --external-sources \
  --severity=warning \
  $( \
    grep -r -E '(^\s*((\#|\!)|(\#\s*\!)|(\!\s*\#))\s*(\/usr(\/local)?)?\/bin\/(env\s+)?(sh|ash|bash|dash|ksh|bats)\b)|(^\s*\#\s+-\*-\s+(sh|ash|bash|dash|ksh|bats)\s+-\*-\s*)|(^\s*\#\s*shellcheck\s+shell=(sh|ash|bash|dash|ksh|bats)\s*)' './' \
    | cut -d: -f1 \
  ) 
```

> **Note**: When including also severity `note` it's **5604** defects reported

> **Note**: Also, `differential-shellcheck` runs in default ShellCheck settings, only including `--external-sources`, but it can be configured to your needs.

**Comment by jamacku** [comment]
> Or something similar. In fact, this may be a better alternative to using github-actions and a differential check. Then we can just get everything actually fixed :smile:

@cevich,  I'm not so familiar with the git pre-commit hook, but wouldn't it block you from pushing until you resolve all issues? If so, users not experienced in the shell would not even be able to propose changes for the first reviews. I think using github-action together with SARIF reports is a better user experience. But git pre-commit hook is undoubtedly an option. I saw a few repositories that use it together with ShellCheck.

**Comment by edsantiago** [comment]
>here is a list of ShellCheck defects

Thank you, that tells me everything I needed to know. This is a no-go from my perspective, sorry.

**Comment by rhatdan** [comment]
Is there an easy way to tone it down?  IE Checks we don't care about?

**Comment by jamacku** [comment]
> Thank you, that tells me everything I needed to know. This is a no-go from my perspective, sorry.

Could you please extend your answer a bit? It's not clear to me. 

From my point of view, it is better to have some linter that wouldn't block new features and changes. At the same time, it would "notify" you about incoming "new" defects. And at the same time, you can work on fixing existing "old" defects.

But I don't want to push you to merge this PR. Feel free to close it. Thanks for the discussion. :+1:

**Comment by jamacku** [comment]
> Is there an easy way to tone it down? IE Checks we don't care about?

Certain checks can be disabled/enabled using [`.shellcheckrc`](https://github.com/koalaman/shellcheck/blob/master/shellcheck.1.md#rc-files)

**Comment by edsantiago** [comment]
Obvious first step is to disable checks on `vendor`. I don't think there's any controversy about that.

Even then, though, there's still a ton of false alarms, errors like "this variable is not defined anywhere". Yes, it is, in the calling environment. Adding lint comments to each of those is IMHO unpleasant, adding clutter without value. The bats and apiv2 tests are special cases, where the individual script files assume a well-defined environment.

Some of the warnings are legitimate, though. I've added a TODO marker for me to explore those later.

**Comment by rhatdan** [comment]
Could you prioritize the ones that you think we should fix?  And then eliminate the ones which are low priority.

**Comment by cevich** [comment]
There's going to be a a lot of noise from `contrib/cirrus/*` as many of those scripts have specific environment requirements (pre-cooked into our VM and contianer images).

**Comment by cevich** [comment]
> block you from pushing until you resolve all issues

Yes, but arguably those are problems that need fixing (or check disablement).

> 899 defects

Oof.

> But git pre-commit hook is undoubtedly an option

The nice thing with this is (when installed) it will cause `make lint` to alter developers before they even create a PR.  In fact, it won't even allow a commit to be made if there are findings (there are ways to disable it).  The only "downside" to it is having to install and maintain yet-another-tool locally.

> SARIF reports

I don't know what that is.  I'm assuming it's some wiz-bang github WebUI element that brings attention to problems.

**Comment by cevich** [comment]
> Could you prioritize the ones that you think we should fix? And then eliminate the ones which are low priority.

This is possible via a "magic" comment.  Unfortunately it will need to be spread around like butter for some things.  For example, it's going to issue "undefined variable" complaints for every `$CIRRUS_<foo>` variable referenced.  So, there's some up-front work required, but after removing the `vendor` directory, it might not be so bad to fix them (I'm hoping).

**Comment by cevich** [comment]
My preference is slightly in favor of updating `.pre-commit-config.yaml` over github-actions because:

* It's optional (if you don't want it, don't install `pre-commit`)
* It's easy for developers to change and/or override the options
* It's easy to understand
* It more portable beyond github
* It doesn't lock the repo. to github
* It's a simpler design

All that said...this use-case is simple enough I'm not strongly opposed to it living in Github action workflow. The differential feature sounds nice, but I'm not entirely convinced we shouldn't simply fix the broken things up-front.

**Comment by vrothberg** [comment]
I am supportive of whatever you folks decide.  False positives are certainly an annoyance.

**Comment by cevich** [comment]
Update: I checked, and we've already got a shellcheck job running for all of `contrib/cirrus` and `hack/get_ci_vm.sh` in https://github.com/containers/podman/blob/main/contrib/cirrus/shellcheck.sh which runs in the `Check Automation` Cirrus-CI task.

Perhaps it would simpler to just extend that to cover other important scripts?  I don't think the number of findings should be excessively large or hard to fix.

**Comment by vrothberg** [comment]
@cevich am I reading it correctly that this PR is redundant?

**Comment by cevich** [comment]
Oh thanks for asking, I forgot about this.  It's somewhat redundant, except for the `/hack*` dir., which might be valuable to check.  That could easily be fixed by [updating this line](https://github.com/containers/podman/blob/main/contrib/cirrus/prebuild.sh#L37).  However...it will likely also mean actually fixing a bunch of hack scripts that are otherwise "working" :confused:  While I certainly won't argue against fixing them, I don't want to do that work :grinning: If we're willing to leave the duct-tape applied to them, I'm okay with closing this.

**Comment by vrothberg** [comment]
@cevich entirely up to you.  I just want to make sure we don't forget about this PR.

**Comment by cevich** [comment]
Okay, let's close this then.  But thank you very much @jamacku for the contribution.  We get very little outside CI/automation help, and it's (clearly) a challenging topic from multiple POVs.  So please don't take my closing this as any reflection on how much we appreciate the help.  Thanks again.

**Pull request closed without merging**
